### ISSUE TITLE
Gem-based theme test from scratch

### ISSUE THREAD

**Issue by DirtyF** [mmistakes/minimal-mistakes#549]
Hi @mmistakes 

I just ran a quick test of your theme from scratch, and there's a few things that cause a non working theme out of the box (related mostly to missing files and config values).
- [x] change the default `minima` theme with `minimal-mistakes-jekyll` in the `_config.yml` (caution, you mentioned `gem` instead of `theme` in your post)
- [x] update url value in config file (or example.com by default  will cause trouble)
- [x] remove default minima files (css, about page) generated with `jekyll new`
- [x] replace default index.html with index.html from example folder shipped in the gem
- [x]  add examples files to your website to add content to display

I guess it's more easy to switch from an existing minimal-theme but I thought I let you know, because there's more than just the `_data` folder to add to your website to make this theme work.

**Comment by mmistakes** [comment]
Thanks for taking it for a spin @DirtyF :smile:

You're right, documentation and example side of things definitely needs to be built out more. Specifically from a `jekyll new` perspective as there's a few things that need to be changed to get things looking right.

The biggie that I left off from my quick blog post last night, was about replacing `index.html` with the [modified version MM](https://github.com/mmistakes/minimal-mistakes/blob/master/index.html) uses. The home page will appear messed up if the one created by `jekyll new` is used.

There's also the layout names that need adjusting `layout: post` or `layout: page` should be `layout: single` in MM-land.

Was thinking of possibly creating some "alias" _layouts to ease the transition, example:

**_layouts/post.html**

```yaml
---
layout: single
---
```

But not sure if that is overkill or not.

Every theme dev does things a little different so still trying to decide on what's the best practice here. Probably stick with what `minima` adopts as it will set the precedent for others to follow...

**Comment by DirtyF** [comment]
@mmistakes GitHub seems to be working on themes for Jekyll, maybe you can ask @broccolini or look at swiss and [Athena](https://github.com/broccolini/athena) themes, because Jekyll'users could benefit from naming convention.

**Comment by mmistakes** [comment]
@DirtyF I've always been partial to how Wordpress names templates.

https://developer.wordpress.org/themes/basics/template-hierarchy/

`default`, `post`, `page` make a lot of sense and are probably there most commonly named layouts. I went with `single` as there was a lot of code overlap that I could DRY up combining into over layout.

**Comment by broccolini** [comment]
Right now `Jekyll new` doesn't play very nicely with using a new gem-based theme, unless it follows a very similar setup to `Minima`. The Jekyll folks are working on improvements that will de-couple this a bit more, but right now it's better to setup a vanilla Jekyll site and then add the theme gem. 

For people making themes this probably means creating a simple Jekyll site that people can clone, with the required config needed for your theme, an `index.html`, and example `post.md` and page files that match your layout options. So that it's as simple as possible for people to get the theme working it's good to [set defaults](https://github.com/broccolini/swiss/blob/master/assets/style.scss#L8) for your config options—for example with my Swiss theme, users have the option of setting a [theme color in the config](https://github.com/broccolini/swiss/blob/master/_config.yml#L30) which uses a different [`scss` file](https://github.com/broccolini/swiss/tree/master/_sass). However, if they don't set that in the config it will use the default `black` theme and just work.

There are some updates coming in [Jekyll 3.3](https://github.com/jekyll/jekyll/issues/5400) that will make theme setup easier, i.e. you'll be able to include an `assets` folder in your theme gem and then add a `styles.css` that imports all your `scss` partials—which means someone using your theme doesn't have to do that setup.

Hope that helps and sorry if it's a bit hard to follow. I'll post a link here when I've updated my docs and setup a vanilla site as an example.

**Comment by mmistakes** [comment]
@broccolini That all makes sense and is pretty much what I do (or plan to do).

I think were things start to fall apart are `_layout` names. Default `jekyll new` expects `default.html`, `post.html`, and `page.html`. And because those are the defaults a good majority of theme devs adopted those names.

Where it gets iffy is switching between themes. If everyone uses different naming errors start creeping up. I use a few different names and document that, but most people don't read docs.

I was thinking that maybe some sort of "alias" layout that uses the default names could be a way around this. I suppose this is a larger question for the community to settle on best practices to keep things as seamless as possible when switching out themes.

**Comment by ashmaroli** [comment]
along the similar lines of using an alias, I think the following would be a great:
- since all Jekyll sites have `posts` and `pages`, we could have something like using the `_config.yml` to define layouts:
```yaml
layouts: 
  - index: home.html
  - pages: page.html
  - posts: post.html
```
and then use this info in the frontmatter:
```yaml
---
title: A post
layout: {{ site.layouts.posts }}
---
```

Individual theme docs can inform users what values to assign to those keys above..

But I dont know enough, if this is just fantasy or even possible right now..

**Comment by mmistakes** [comment]
@ashmaroli That could potentially work but requires users to update all of their content to have `layout: {{ site.layouts.posts }}`.

Would be a fine solution going forward, if you could get all theme devs on the same page. But I'm guessing that's going to be a hard sell unless it's baked into the default `jekyll new` behavior.

I'm not too worried about either way as I encourage users to use YAML Front Matter defaults to set layouts and keep it out of the actual content.

The problem is more for `jekyll new` users or someone coming from another theme who don't bother reading the docs on changing the layouts and get an error with `jekyll serve`.

**Comment by ashmaroli** [comment]
Thank you for the feedback.. I'll bring this up in the main repo once all the dust surrounding theme gems and new directory structures clear out a bit..

**Comment by mmistakes** [comment]
Sounds good @ashmaroli. Really excited for the `assets` theme folder PR to release. All of the new gem related theme stuff is such an improvement!

**Comment by fa-ribeiro** [comment]
Hi all,

I just tested the new gem based minimal-mistakes theme and would like to share my experience, hoping it can be useful.

Here are the steps I followed:

#### 1. Create an empty dir

    mkdir mm
    cd mm

#### 2. Create the `Gemfile`

Create a `Gemfile` as described [here](https://mmistakes.github.io/minimal-mistakes/jekyll/gemified-theme-alpha/):

    source "https://rubygems.org"
    
    gem "jekyll", :git => "https://github.com/jekyll/jekyll.git"
    gem "minimal-mistakes-jekyll", :git => "https://github.com/mmistakes/minimal-mistakes.git", :branch => "feature/theme-gem"
    
    group :jekyll_plugins do
      # gem "jekyll-archives"
      gem "jekyll-paginate"
      gem "jekyll-sitemap"
      gem "jekyll-gist"
      gem "jekyll-feed"
      gem "jemoji"
    end

#### 3. Create the `index.html` file

Create the `index.html` file as explained [here](https://mmistakes.github.io/minimal-mistakes/jekyll/gemified-theme-alpha/). I just copied the content from [here](https://github.com/mmistakes/minimal-mistakes/blob/master/index.html):

    ---
    layout: archive
    author_profile: true
    ---
    
    {% include base_path %}
    
    <h3 class="archive__subtitle">{{ site.data.ui-text[site.locale].recent_posts | default: "Recent Posts" }}</h3>
    
    {% for post in paginator.posts %}
      {% include archive-single.html %}
    {% endfor %}
    
    {% include paginator.html %}

#### 4. Create the `_config.yml` file

Create a standard minimal-mistakes `_config.yml` file. Again, I took it from [here](https://raw.githubusercontent.com/mmistakes/minimal-mistakes/feature/theme-gem/example/_config.yml).

#### 5. Go!

Then I just run:

    bundle install

followed by:

    bundle exec jekyll serve

Et voilá, a bare minimum blog working like a charm :-)

[Here](https://github.com/mmistakes/minimal-mistakes/files/490553/mm.zip) you can find the 3 files I use in this very minimal setup.

**Comment by mmistakes** [comment]
That is antastic @fa-ribeiro, thanks for taking it for a spin! That's exactly how I was envisioning a fresh install of the theme to go and glad to hear it worked out.

Hopefully I can ship soon. All that's in my way is waiting on Jekyll core and `github-page` to merge in a few things and :confetti_ball:

**Comment by mmistakes** [comment]
Also I'd like to add that I pushed the MM gem so it can be referenced directly in `Gemfile` instead of via its GH branch.

```diff
-  gem "minimal-mistakes-jekyll", :git => "https://github.com/mmistakes/minimal-mistakes.git", :branch => "feature/theme-gem"
+  gem "minimal-mistakes-jekyll", "~> 4.0', '>= 4.0.1"
```

**Comment by bafonso** [comment]
When I installed this theme coming from the so simple I did not get a _layout directory or any files in it...

**Comment by mmistakes** [comment]
@bafonso That's correct. When you use the theme as a gem (or remote theme) you won't "see" any of the layout, include, asset files in your repo. Those are pulled in directly from the gem.

See Jekyll's documentation on how they work for more info.

https://jekyllrb.com/docs/themes/#understanding-gem-based-themes
